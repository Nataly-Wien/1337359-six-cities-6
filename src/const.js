export const NOT_AUTHORIZED_USERNAME = `Sign in`;
export const EMPTY_AVATAR_URL = `../img/avatar.svg`;
export const HOTEL_RATING_VALUES = [`perfect`, `good`, `not bad`, `badly`, `terribly`];
export const CITIES = [`Paris`, `Cologne`, `Brussels`, `Amsterdam`, `Hamburg`, `Dusseldorf`];
export const REVIEWS_AMOUNT = 10;

export const ReviewsLength = {
  MIN: 50,
  MAX: 300,
};

export const SORTS = [
  {
    type: `Popular`,
    rule: (a, b) => (a - b) * 0,
  },
  {
    type: `Price`,
    rule: (a, b) => a.price - b.price,
  },
  {
    type: `Price: high to low`,
    rule: (a, b) => b.price - a.price,
  },
  {
    type: `Top rated first`,
    rule: (a, b) => b.rating - a.rating,
  }
];

export const Routes = {
  HOME: `/`,
  LOGIN: `/login`,
  FAVORITES: `/favorites`,
  ROOM: `/offer/:id`,
  NOT_FOUND: `/404`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const emptyUser = {
  id: -1,
  email: ``,
  password: ``,
  name: ``,
  avatarUrl: EMPTY_AVATAR_URL,
  isPro: false,
};

export const Url = {
  HOTELS: `/hotels`,
  LOGIN: `/login`,
  LOGOUT: `/logout`,
  NEAR: `/nearby`,
  COMMENTS: `/comments`,
  FAVORITES: `/favorite`,
};

export const LeafletConst = {
  TITLE_LAYER: `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`,
  ATTRIBUTION: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
};

export const MAP_ICON = {
  iconUrl: `img/pin.svg`,
  iconSize: [27, 39],
};

export const MAP_ICON_ACTIVE = {
  iconUrl: `img/pin-active.svg`,
  iconSize: [27, 39],
};

export const Types = {
  MAIN_PAGE: `MAIN_PAGE`,
  NOT_MAIN_PAGE: `NOT_MAIN_PAGE`,
  FAVORITES_PAGE: `FAVORITES_PAGE`,
  ROOM_PAGE: `ROOM_PAGE`,
  CARD: `CARD`,
};

export const CardTypes = {
  MAIN_PAGE: {
    articleClassName: `cities__place-card`,
    imgWrapperClassName: `cities__image-wrapper`,
    cardInfoClassName: ``,
    hasPremiumMark: true,
    imgWidth: `260`,
    imgHeight: `200`,
  },
  FAVORITES_PAGE: {
    articleClassName: `favorites__card`,
    imgWrapperClassName: `favorites__image-wrapper`,
    cardInfoClassName: `favorites__card-info`,
    hasPremiumMark: false,
    imgWidth: `150`,
    imgHeight: `110`,
  },
  ROOM_PAGE: {
    articleClassName: `near-places__card`,
    imgWrapperClassName: `near-places__image-wrapper`,
    cardInfoClassName: ``,
    hasPremiumMark: false,
    imgWidth: `260`,
    imgHeight: `200`,
  },
};

export const OfferTypes = {
  MAIN_PAGE: {
    offerClassName: `cities__places-list places__list tabs__content`,
  },
  FAVORITES_PAGE: {
    offerClassName: `favorites__places`,
  },
  ROOM_PAGE: {
    offerClassName: `near-places__list places__list`,
  },
};

export const FavoriteMarkTypes = {
  CARD: {
    buttonClassTerm: `place-card`,
    imgWidth: `18`,
    imgHeight: `19`,
  },
  ROOM_PAGE: {
    buttonClassTerm: `property`,
    imgWidth: `31`,
    imgHeight: `33`,
  },
};

export const PremiumMarkTypes = {
  CARD: {
    markClassName: `place-card__mark`,
  },
  ROOM_PAGE: {
    markClassName: `property__mark`,
  },
};

export const HeaderTypes = {
  MAIN_PAGE: {
    logoClassName: ` header__logo-link--active`,
  },
  NOT_MAIN_PAGE: {
    logoClassName: ``,
  },
};
