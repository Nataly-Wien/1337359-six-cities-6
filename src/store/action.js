export const ActionType = {
  SET_CITY: `offers/setCity`,
  SET_SORT: `offers/setSort`,

  LOAD_HOTELS: `data/loadHotels`,
  FAILURE_LOAD_HOTELS: `data/failureLoadHotels`,

  REQUEST_CURRENT_HOTEL: `data/requestCurrentHotel`,
  LOAD_CURRENT_HOTEL: `data/loadCurrentHotel`,
  FAILURE_CURRENT_HOTEL: `data/failureCurrentHotel`,

  REQUEST_NEAR_HOTELS: `data/requestNearHotel`,
  LOAD_NEAR_HOTELS: `data/loadNearHotels`,
  FAILURE_NEAR_HOTELS: `data/failureNearHotels`,

  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  FAILURE_AUTHORIZATION: `user/failureAuthorization`,
  SET_USER: `user/setUser`,
  LOGOUT: `user/logout`,

  REQUEST_COMMENTS: `data/requestComments`,
  LOAD_COMMENTS: `data/loadComments`,
  FAILURE_COMMENTS: `data/failureComments`,

  REQUEST_FAVORITES: `data/requestFavorites`,
  LOAD_FAVORITES: `data/loadFavorites`,
  FAILURE_FAVORITES: `data/failureFavorites`,
  REPLACE_HOTEL: `data/replaceHotel`,

  RESET_CURRENT_OFFER: `data/resetCurrentOffer`,

  // SET_ERROR_STATUS: `data/setErrorStatus`,
};

export const ActionCreator = {
  setCity: (payload) => ({
    type: ActionType.SET_CITY,
    payload,
  }),
  setSort: (payload) => ({
    type: ActionType.SET_SORT,
    payload,
  }),

  loadHotels: (payload) => ({
    type: ActionType.LOAD_HOTELS,
    payload,
  }),
  failureLoadHotels: () => ({
    type: ActionType.FAILURE_LOAD_HOTELS,
  }),

  requestCurrentHotel: () => ({
    type: ActionType.REQUEST_CURRENT_HOTEL,
  }),
  loadCurrentHotel: (payload) => ({
    type: ActionType.LOAD_CURRENT_HOTEL,
    payload,
  }),
  failureCurrentHotel: () => ({
    type: ActionType.FAILURE_CURRENT_HOTEL,
  }),

  requestNearHotels: () => ({
    type: ActionType.REQUEST_NEAR_HOTELS,
  }),
  loadNearHotels: (payload) => ({
    type: ActionType.LOAD_NEAR_HOTELS,
    payload,
  }),
  failureNearHotels: () => ({
    type: ActionType.FAILURE_NEAR_HOTELS,
  }),

  requestComments: () => ({
    type: ActionType.REQUEST_COMMENTS,
  }),
  loadComments: (payload) => ({
    type: ActionType.LOAD_COMMENTS,
    payload,
  }),
  failureComments: () => ({
    type: ActionType.FAILURE_COMMENTS,
  }),

  requestFavorites: () => ({
    type: ActionType.REQUEST_FAVORITES,
  }),
  loadFavorites: (payload) => ({
    type: ActionType.LOAD_FAVORITES,
    payload,
  }),
  failureFavorites: () => ({
    type: ActionType.FAILURE_FAVORITES,
  }),
  replaceHotel: (payload) => ({
    type: ActionType.REPLACE_HOTEL,
    payload,
  }),

  resetCurrentOffer: () => ({
    type: ActionType.RESET_CURRENT_OFFER,
  }),

  requiredAuthorization: (payload) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload,
  }),
  logout: () => ({
    type: ActionType.LOGOUT,
  }),
  setUser: (payload) => ({
    type: ActionType.SET_USER,
    payload,
  }),

  setErrorStatus: (payload) => ({
    type: ActionType.SET_ERROR_STATUS,
    payload,
  }),
  // failureAuthorization: () => ({
  //   type: ActionType.FAILURE_AUTHORIZATION,
  // }),
};
