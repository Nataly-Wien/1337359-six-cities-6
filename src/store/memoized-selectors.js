import {createSelector} from 'reselect';
import {getCity, getSort} from './offers/selectors';
import {getHotels} from './data/selectors';
import {SORTS} from '../const';

export const getFilteredSortedHotels = createSelector([getHotels, getCity, getSort], (hotels, city, sort) =>
  hotels.filter((item) => item.city.name === city).sort(SORTS[sort].rule)
);
