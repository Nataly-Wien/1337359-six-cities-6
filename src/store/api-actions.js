import {ActionCreator} from './action';
import {AuthorizationStatus, Url} from '../const';
import {adaptHotelToClient, adaptUserToClient} from '../common';

export const fetchHotels = () => (next, _getState, api) => (
  api.get(Url.HOTELS)
    .then(({data}) => next(ActionCreator.loadHotels(data.map(adaptHotelToClient))))
);

export const checkAuth = () => (next, _getState, api) => (api.get(Url.LOGIN)
  .then(({data}) => next(ActionCreator.requiredAuthorization({status: AuthorizationStatus.AUTH, user: adaptUserToClient(data)})))
  .catch(() => { })
);

export const login = ({login: email, password}) => (next, _getState, api) => (api.post(Url.LOGIN, {email, password})
  .then(({data}) => next(ActionCreator.requiredAuthorization({status: AuthorizationStatus.AUTH, user: adaptUserToClient(data)})))
  .catch(() => {

  })
);

export const logout = () => (next, _getState, api) => (api.get(Url.LOGOUT)
  .then(() => next(ActionCreator.logout()))
);
