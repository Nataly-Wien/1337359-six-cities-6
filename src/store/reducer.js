import {ActionType} from './action';
import {CITIES, AuthorizationStatus, emptyUser} from '../const';

const initialState = {
  city: CITIES[0],
  sort: 0,
  hotels: [],
  currentHotel: {},
  nearHotels: [],
  comments: [],
  favorites: [],
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  user: emptyUser,
  isHotelsLoaded: false,
  isCurrentLoading: false,
  isNearLoading: false,
  isCommentsLoading: false,
  isFavoritesLoading: false,
  isLoadingError: false,
  errorStatus: 0,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_CITY:
      return {
        ...state,
        city: action.payload,
      };

    case ActionType.SET_SORT:
      return {
        ...state,
        sort: action.payload,
      };

    case ActionType.LOAD_HOTELS:
      return {
        ...state,
        hotels: action.payload,
        isHotelsLoaded: true,
      };

    case ActionType.FAILURE_LOAD_HOTELS:
      return {
        ...state,
        isHotelsLoaded: true,
        isLoadingError: true,
      };

    case ActionType.REQUEST_CURRENT_HOTEL:
      return {
        ...state,
        isCurrentLoading: true,
        isLoadingError: false,
      };

    case ActionType.LOAD_CURRENT_HOTEL:
      return {
        ...state,
        currentHotel: action.payload,
        isCurrentLoading: false,
      };

    case ActionType.FAILURE_CURRENT_HOTEL:
      return {
        ...state,
        isCurrentLoading: false,
        isLoadingError: true
      };

    case ActionType.REQUEST_NEAR_HOTELS:
      return {
        ...state,
        isNearLoading: true,
      };

    case ActionType.LOAD_NEAR_HOTELS:
      return {
        ...state,
        nearHotels: action.payload,
        isNearLoading: false,
      };

    case ActionType.FAILURE_NEAR_HOTELS:
      return {
        ...state,
        isNearLoading: false,
        isLoadingError: false,
      };

    case ActionType.REQUEST_COMMENTS:
      return {
        ...state,
        isCommentsLoading: true,
      };

    case ActionType.LOAD_COMMENTS:
      return {
        ...state,
        comments: action.payload,
        isCommentsLoading: false,
      };

    case ActionType.FAILURE_COMMENTS:
      return {
        ...state,
        isCommentsLoading: false,
        isLoadingError: false,
      };

    case ActionType.REQUEST_FAVORITES:
      return {
        ...state,
        isFavoritesLoading: true,
        isLoadingError: false,
      };

    case ActionType.LOAD_FAVORITES:
      return {
        ...state,
        favorites: action.payload,
        isFavoritesLoading: false,
      };

    case ActionType.FAILURE_FAVORITES:
      return {
        ...state,
        isFavoritesLoading: false,
        isLoadingError: true,
      };

    case ActionType.REPLACE_HOTEL:
      return {
        ...state,
        hotels: state.hotels.map((item) => item.id === action.payload.id ? action.payload : item),
        currentHotel: state.currentHotel.id === action.payload.id ? action.payload : state.currentHotel,
        favorites: state.favorites.map((item) => item.id === action.payload.id ? action.payload : item),
      };

    case ActionType.RESET_CURRENT_OFFER:
      return {
        ...state,
        currentHotel: {},
        nearHotels: [],
        comments: [],
      };

    case ActionType.REQUIRED_AUTHORIZATION:
      return {
        ...state,
        authorizationStatus: action.payload,
      };

    case ActionType.SET_USER:
      return {
        ...state,
        user: action.payload,
      };

    case ActionType.LOGOUT:
      return {
        ...state,
        authorizationStatus: AuthorizationStatus.NO_AUTH,
        user: emptyUser,
      };

    case ActionType.SET_ERROR_STATUS:
      return {
        ...state,
        errorStatus: action.payload,
      };
  }

  // case ActionType.FAILURE_AUTHORIZATION:
  //   return {
  //     ...state,
  //     isLoadingError: true,
  //   };

  return state;
};

export {reducer};
