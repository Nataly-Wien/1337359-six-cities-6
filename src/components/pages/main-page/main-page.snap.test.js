import React from 'react';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';
import {render} from '@testing-library/react';
import thunk from 'redux-thunk';
import * as APIActions from '../../../store/api-actions';
import MainPage from './main-page';
import {AuthorizationStatus} from '../../../const';

const history = createMemoryHistory();

jest.mock(`../../../store/api-actions.js`);
APIActions.fetchHotels = () => (dispatch) => dispatch(jest.fn());

const middleware = [thunk];
const mockStore = configureStore(middleware);
const store = mockStore({
  OFFERS: {
    city: `Paris`,
    sort: 0,
  },
  DATA: {
    hotels: [{
      id: 5,
      city: {
        name: `Amsterdam`,
        location: {
          latitude: 52.37454,
          longitude: 4.897976,
          zoom: 12
        },
      },
      previewImage: `img/img.jpg`,
      images: [
        `img/img1.jpg`,
        `img/img2.jpg`,
        `img/img3.jpg`,
        `img/img4.jpg`,
        `img/img5.jpg`,
        `img/img6.jpg`,
      ],
      title: `Cityden Residences Museum Square`,
      description: `In the center of Paris, Citadines Saint-Germain-des-PrÃ©s Paris is a 10-minute walk from Notre-Dame Cathedral and 293 m from Saint-Michel Metro Station. It features a fitness room and free Wi-Fi access.`,
      location: {
        latitude: 52.3909553943508,
        longitude: 4.85309666406198,
        zoom: 16
      },
      isFavorite: true,
      isPremium: true,
      rating: 4.1,
      type: `apartment`,
      bedrooms: 2,
      maxAdults: 4,
      price: 210,
      goods: [
        `Air conditioning`,
        `Private Bathroom`,
        `Dishwasher`,
        `Flat - screen TV`,
        `Coffee machine`,
        `Free WiFi`,
      ],
      host: {
        id: 10,
        name: `Mick`,
        isPro: true,
        avatarUrl: `img/avatar.jpg`,
      },
    }],
    isLoadingError: false,
    isHotelsLoading: false,
  },
  USER: {
    authorizationStatus: AuthorizationStatus.AUTH,
    user: {
      avatarUrl: `img/avatar.jpg`,
      email: `Oliver.conner@gmail.com`,
      id: 1,
      name: `Oliver.conner`,
    },
  },
});

it(`Should favorites page render correctly`, () => {
  const {container} = render(<Provider store={store}>
    <Router history={history}>
      <MainPage />
    </Router>
  </Provider>);

  expect(container).toMatchSnapshot();
});
