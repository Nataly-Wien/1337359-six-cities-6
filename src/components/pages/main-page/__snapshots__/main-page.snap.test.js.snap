// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should favorites page render correctly 1`] = `
<div>
  <div
    class="page page--gray page--main"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo-link header__logo-link--active"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
                style="display: flex;"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  href="/favorites"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                    style="background-image: url(img/avatar.jpg);"
                  />
                  <span
                    class="header__user-name user__name"
                  >
                    Oliver.conner@gmail.com
                  </span>
                </a>
                <button
                  class="button"
                  style="margin-left: 10px; width: 18px; height: 18px; background-image: url(../img/logout.svg);"
                />
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page__main page__main--index page__main--index-empty"
    >
      <h1
        class="visually-hidden"
      >
        Cities
      </h1>
      <div
        class="tabs"
      >
        <section
          class="locations container"
        >
          <ul
            class="locations__list tabs__list"
          >
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item  tabs__item--active"
                href="#"
              >
                <span>
                  Paris
                </span>
              </a>
            </li>
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item "
                href="#"
              >
                <span>
                  Cologne
                </span>
              </a>
            </li>
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item "
                href="#"
              >
                <span>
                  Brussels
                </span>
              </a>
            </li>
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item "
                href="#"
              >
                <span>
                  Amsterdam
                </span>
              </a>
            </li>
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item "
                href="#"
              >
                <span>
                  Hamburg
                </span>
              </a>
            </li>
            <li
              class="locations__item"
            >
              <a
                class="locations__item-link tabs__item "
                href="#"
              >
                <span>
                  Dusseldorf
                </span>
              </a>
            </li>
          </ul>
        </section>
      </div>
      <div
        class="cities"
      >
        <div
          class="cities__places-container cities__places-container--empty container"
        >
          <section
            class="cities__no-places"
          >
            <div
              class="cities__status-wrapper tabs__content"
            >
              <b
                class="cities__status"
              >
                No places to stay available
              </b>
              <p
                class="cities__status-description"
              >
                We could not find any property available at the moment in 
                Paris
              </p>
            </div>
          </section>
          <div
            class="cities__right-section"
          />
        </div>
      </div>
    </main>
  </div>
</div>
`;
